src/data/program_configuration.js

```diff
@@ -246,8 +322,131 @@ class CompositeExpressionBinder<T> implements Binder<T> {
     }
 
     setUniforms(context: Context, program: Program, globals: GlobalProperties) {
-        context.gl.uniform1f(program.uniforms[`a_${this.name}_t`], this.interpolationFactor(globals.zoom));
+        context.gl.uniform1f(program.uniforms[`a_${this.names[0]}_t`], this.interpolationFactor(globals.zoom));
     }
+
+    setTileSpecificUniforms() {}
+}
+
+class PatternCompositeExpressionBinder<T> implements Binder<T> {
+    expression: CompositeExpression;
+    names: Array<string>;
+    type: string;
+    useIntegerZoom: boolean;
+    zoom: number;
+    statistics: { max: number };
+
+    zoomInPaintVertexArray: StructArray;
+    zoomOutPaintVertexArray: StructArray;
+    zoomInPaintVertexBuffer: ?VertexBuffer;
+    zoomOutPaintVertexBuffer: ?VertexBuffer;
+    paintVertexAttributes: Array<StructArrayMember>;
+
+    constructor(expression: CompositeExpression, names: Array<string>, type: string, useIntegerZoom: boolean, zoom: number, PaintVertexArray: Class<StructArray>) {
+
+        this.expression = expression;
+        this.names = names;
+        this.type = type;
+        this.useIntegerZoom = useIntegerZoom;
+        this.zoom = zoom;
+        this.statistics = { max: -Infinity };
+
+        this.paintVertexAttributes = names.map((name) =>
+            ({
+                name: `a_${name}`,
+                type: 'Float32',
+                components: 4,
+                offset: 0
+            })
+        );
+
+        this.zoomInPaintVertexArray = new PaintVertexArray();
+        this.zoomOutPaintVertexArray = new PaintVertexArray();
+    }
+
+    defines() {
+        return [];
+    }
+
+    populatePaintArray(length: number, feature: Feature, imagePositions: ?{[string]: ImagePosition}) {
```

Comment:
I'd rather have `imagePositions` be required parameter, both here and on the `Binder` interface, and a parameter defaulting to `{}` in `ProgramConfigurationSet#populatePaintArray()`